### PR TITLE
[Mellanox]enhance the initialization flow of sfp part of new platform api

### PR DESCRIPTION
**Pull request by stephenxs**
**- What I did**
Port [mlnx-sfpd init flow enhancement #3294](https://github.com/Azure/sonic-buildimage/pull/3294) to sfp_event.
On Mellanox devices sfp module insertion/removal is monitored via sdk notification.
Originally this is done in mlnx-sfpd task which is running in syncd.
In new platform api it is done directly in the chassis.get_change_event which runs in pmon and talks to sdk directly so that the optimization should be ported to new plaform api.

**- How I did it**
Port to new platform api

**- How to verify it**
Test sfp insert/remove flow.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 1]
```diff
@@ -24,54 +24,108 @@
...
+                if retry >= self.SX_OPEN_RETRIES:
+                    raise RuntimeError("SDK daemon failed to start after {} retries and {} seconds waiting, exiting..."
+                        .format(retry, self.SX_OPEN_TIMEOUT * self.SX_OPEN_RETRIES))
+                else:
+                    logger.log_info("SDK daemon not started yet, retry {} times".format(retry))
+                    retry = retry + 1
+                    time.sleep(self.SX_OPEN_TIMEOUT)
+
+            # to make sure SDK daemon has started
+            time.sleep(self.SX_OPEN_TIMEOUT)
```
Comment (andriymoroz-mlnx) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: this is probably not needed...

Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: Hi,
This is a workaround for an issue. Kebo and I have observed the phenomenon where sx_api_open can fail even after sdk_ready existing. It is very difficult to reproduce with SX_OPEN_TIMEOUT as 5 seconds but easy as 0.5 second. It's likely to be a race condition. 
You remind me that I should raise an issue for it.

Comment (nazariig) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: @stephenxs It is very difficult to reproduce with SX_OPEN_TIMEOUT being 5 seconds but easy with ...
IMHO, If it's still reproducible then we should treat it as an issue

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 2]
```diff
@@ -24,54 +24,108 @@
...
+        try:
+            # Wait for SDK daemon to be started with detect the sdk_ready file
+            retry = 0
+            while not os.path.exists(SDK_DAEMON_READY_FILE):  
+                if retry >= self.SX_OPEN_RETRIES:
+                    raise RuntimeError("SDK daemon failed to start after {} retries and {} seconds waiting, exiting..."
+                        .format(retry, self.SX_OPEN_TIMEOUT * self.SX_OPEN_RETRIES))
+                else:
+                    logger.log_info("SDK daemon not started yet, retry {} times".format(retry))
+                    retry = retry + 1
```
Comment (nazariig) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: @stephenxs For consistency please replace with '+='

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py [hunk 3]
```diff
@@ -24,54 +24,108 @@
...
+                            swid_cnt_p = None
+                            break
+                        else:
+                            logger.log_info("switch not created yet, swid_cnt {}, retry {} times and wait for {} seconds"
+                                .format(swid_cnt, retry, self.SX_OPEN_TIMEOUT * retry))
+                    else:
+                        raise RuntimeError("sx_api_port_swid_list_get fail with rc {}, retry {} times and wait for {} seconds".
+                            format(rc, retry, self.SX_OPEN_TIMEOUT * retry))
+
+                    retry = retry + 1
```
Comment (nazariig) [inline platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: @stephenxs For consistency please replace with '+='

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py (hunk 1 above)
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: > @stephenxs It is very difficult to reproduce with SX_OPEN_TIMEOUT being 5 seconds but easy with ...
> IMHO, If it's still reproducible then we should treat it as an issue

Yes, it is. An issue will be raised. But I think before it is fixed a workaround is required to keep things work.

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py (hunk 3 above)
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: fixed.

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py (hunk 2 above)
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: fixed.

File: platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py (hunk 1 above)
Comment (stephenxs) [reply platform/mellanox/mlnx-platform-api/sonic_platform/sfp_event.py]: This issue doesn't reproduce since Sept's images. So this statement is removed.

**Pull request merged**